### PR TITLE
Issue #19: Updated run_config.py to stop premature phase advancement.

### PR DESCRIPTION
**Pull request by carolinedav**

### CODE REVIEW

File: run_brer/run_config.py [hunk 1]
```diff
@@ -369,12 +369,15 @@ def __train(self, tpr_file=None, **kwargs):
             current_alpha = context.potentials[i].alpha
             # noinspection PyUnresolvedReferences
             current_target = context.potentials[i].target
+            current_stop_called = context.potentials[i].stop_called
```
Comment (eirrgang) [inline run_brer/run_config.py:372]:
```suggestion
            current_stop_called = getattr(context.potentials[i], 'stop_called', None)
```

File: run_brer/run_config.py [hunk 2]
```diff
@@ -369,12 +369,15 @@ def __train(self, tpr_file=None, **kwargs):
...
 
             self.run_data.set(name=current_name, alpha=current_alpha)
             self.run_data.set(name=current_name, target=current_target)
-            self._logger.info("Plugin {}: alpha = {}, target = {}".format(current_name,
-                                                                          current_alpha,
-                                                                          current_target))
+            self.run_data.set(name=current_stop_called, stop_called=current_stop_called)
```
Comment (eirrgang) [inline run_brer/run_config.py:376]: I'm not sure you can do this without updating the lists of named data. I think that's in the `run_data.py` module.

File: run_brer/run_config.py [hunk 3]
```diff
@@ -369,12 +369,15 @@ def __train(self, tpr_file=None, **kwargs):
...
+            self._logger.info("Plugin {}: alpha = {}, target = {}, stop_called = {}".format(current_name,
```
Comment (eirrgang) [inline run_brer/run_config.py:377]:
```suggestion
            message = f'Plugin {current_name}: alpha = {current_alpha}, target = {current_target}'
            if current_stop_called is not None:
                message += f', stop_called = {current_stop_called}'
            self._logger.info(message)
```

File: run_brer/run_config.py [hunk 4]
```diff
@@ -408,9 +411,11 @@ def __converge(self, tpr_file=None, **kwargs):
         for name in self.__names:
             current_alpha = self.run_data.get('alpha', name=name)
             current_target = self.run_data.get('target', name=name)
-            self._logger.info("Plugin {}: alpha = {}, target = {}".format(name,
-                                                                          current_alpha,
-                                                                          current_target))
+            current_stop_called = self.run_data.get('stop_called', name=name)
```
Comment (eirrgang) [inline run_brer/run_config.py:414]: Oh. Oops. Didn't notice this before. "stop_called" is only available on the LinearStopRestraint potential, which is only used in the convergence phase. We will need different checks in the training and production phases.

**Review by eirrgang — CHANGES_REQUESTED** [review]
Let's make the "stop_called" access robust for older versions of the plugin.

We should only be looking for "stop_called" in the convergence phase, where it is used.

File: run_brer/run_config.py (hunk 1 above)
Comment (carolinedav) [reply run_brer/run_config.py:372]: Resolving in commit

File: run_brer/run_config.py (hunk 2 above)
Comment (carolinedav) [reply run_brer/run_config.py:376]: Pausing on this issue to examine `run_data.py`

File: run_brer/run_config.py (hunk 3 above)
Comment (carolinedav) [reply run_brer/run_config.py:377]: Resolving in commit

Comment (carolinedav) [reply run_brer/run_config.py:377]: Am I removing
```python
                                                                                           current_alpha,
                                                                                           current_target,
                                                                                            current_stop_called))
```
after ` self._logger.info(message)`?

File: run_brer/run_config.py (hunk 4 above)
Comment (carolinedav) [reply run_brer/run_config.py:414]: Removed `current_stop_called` from production but left in convergence function

File: run_brer/run_config.py (hunk 3 above)
Comment (eirrgang) [reply run_brer/run_config.py]: Right. I just couldn't figure out how to make a suggested change apply to multiple lines.

File: run_brer/run_config.py (hunk 2 above)
Comment (eirrgang) [reply run_brer/run_config.py]: I'm looking into it.

**Comment by eirrgang** [comment]
Do you think it is important that `stop_called` is restored after a re-start?

I'm thinking that it might not be worth the added complexity. In a single-simulation Python process (which all of ours currently are), if the convergence phase has `stop_called`, then it will be considered complete and the phase will advance (with this patch).

In the hypothetical case where there is an error before this happens, or if we return to using multiple simulations per Python job, if some simulations have converged while others have not when the job is interrupted, convergence should be re-detected no later than the next converged simulation would have been detected anyway.

In other words, I think we can skip the part of this change that would add `stop_called` as a parameter for the `ConvergencePluginConfig`. Yes?

**Review by eirrgang — APPROVED** [review]
